src/App-context-types.spec.ts

```diff
@@ -0,0 +1,78 @@
+import sinon from 'sinon';
+import rewiremock from 'rewiremock';
+import { mergeOverrides, Override } from './test-helpers';
+import { Receiver, ReceiverEvent } from './types';
+import App from './App';
+
+type IfAny<T, Y, N> = 0 extends (1 & T) ? Y : N;
+interface valid { valid: boolean }
+interface GlobalContext { globalContextKey: number }
+interface MiddlewareContext { middleWareContextKey: number }
+
+// Loading the system under test using overrides
+async function importApp(
+  overrides: Override = mergeOverrides(withNoopAppMetadata(), withNoopWebClient()),
+): Promise<typeof import('./App').default> {
+  return (await rewiremock.module(() => import('./App'), overrides)).default;
+}
+
+class FakeReceiver implements Receiver {
+  private bolt: App | undefined;
+
+  public init = (bolt: App) => {
+    this.bolt = bolt;
+  };
+
+  public start = sinon.fake((...params: any[]): Promise<unknown> => Promise.resolve([...params]));
+
+  public stop = sinon.fake((...params: any[]): Promise<unknown> => Promise.resolve([...params]));
+
+  public async sendEvent(event: ReceiverEvent): Promise<void> {
+    return this.bolt?.processEvent(event);
+  }
+}
+
+const noopAuthorize = () => Promise.resolve({});
+const receiver = new FakeReceiver();
+
+describe('context typing', () => {
+  it('should pass global context to message middleware', async () => {
+
+    const MockApp = await importApp();
+    const app = new MockApp<GlobalContext>({ receiver, authorize: noopAuthorize });
+
+    app.message('hello', async ({ context }) => {
+      // If the globalContextKey in the context is not explicitly typed, then it will be 'any'
+      // The IfAny check will then set 'check' to 'never', causing the check.valid call later to fail
```

Comment:
This is a bit tricky but as long as the `check.valid = true;` statement fails in the compilation phase, it'd be fine.